python/ray/_raylet.pyx

```diff
@@ -0,0 +1,647 @@
+        for i in range(num_args):
+            count = task_spec.ArgIdCount(i)
+            if count > 0:
+                assert count == 1
+                arg_list.append(ObjectID.from_native(task_spec.ArgId(i, 0)))
+            else:
+                serialized_str = task_spec.ArgVal(i)[:task_spec.ArgValLength(i)]
+                obj = pickle.loads(serialized_str)
+                arg_list.append(obj)
+        return arg_list
+
+    def returns(self):
+        """Return the object IDs for the return values of the task."""
+        cdef RayletTaskSpecification *task_spec = self.task_spec.get()
+        return_id_list = []
+        for i in range(task_spec.NumReturns()):
+            return_id_list.append(ObjectID.from_native(task_spec.ReturnId(i)))
+        return return_id_list
+
+    def required_resources(self):
+        """Return the resource vector of the task."""
+        cdef:
+            unordered_map[c_string, double] resource_map = self.task_spec.get().GetRequiredResources().GetResourceMap()
+            c_string resource_name
+            double resource_value
+            unordered_map[c_string, double].iterator iterator = resource_map.begin()
+
+        required_resources = {}
+        while iterator != resource_map.end():
+            resource_name = dereference(iterator).first
+            py_resource_name = str(resource_name)  # bytes for Py2, unicode for Py3
+            resource_value = dereference(iterator).second
+            required_resources[py_resource_name] = resource_value
+            postincrement(iterator)
+        return required_resources
+
+    def actor_creation_id(self):
+        """Return the actor creation ID for the task."""
+        return ActorID.from_native(self.task_spec.get().ActorCreationId())
+
+    def actor_creation_dummy_object_id(self):
+        """Return the actor creation dummy object ID for the task."""
+        return ObjectID.from_native(self.task_spec.get().ActorCreationDummyObjectId())
+
+    def actor_id(self):
+        """Return the actor ID for this task."""
+        return ActorID.from_native(self.task_spec.get().ActorId())
+
+    def actor_counter(self):
+        """Return the actor counter for this task."""
+        return self.task_spec.get().ActorCounter()
+
+
+cdef class RayletClient:
+    cdef unique_ptr[CRayletClient] client
+    def __cinit__(self, raylet_socket,
+                  ClientID client_id,
+                  c_bool is_worker,
+                  DriverID driver_id):
+        # We have known that we are using Python, so just skip the language parameter.
+        self.client.reset(new CRayletClient(raylet_socket.encode("ascii"), client_id.data,
+                          is_worker, driver_id.data, LANGUAGE_PYTHON))
+
+    def disconnect(self):
+        check_status(self.client.get().Disconnect())
+
+    def submit_task(self, Task task_spec):
+        check_status(self.client.get().SubmitTask(task_spec.execution_dependencies.get()[0], task_spec.task_spec.get()[0]))
```

Comment:
Instead of using `[0]`, shouldn't we dereference the pointers? Since `[0]` gives the impression that it is an array.